aptos-move/aptos-gas/src/transaction.rs

```diff
@@ -73,9 +71,88 @@ crate::params::define_gas_parameters!(
             "gas_unit_scaling_factor",
             1000
         ],
+        // Gas Parameters for reading data from storage.
+        [load_data_base: InternalGas, "load_data.base", 1],
+        [
+            load_data_per_byte: InternalGasPerByte,
+            "load_data.per_byte",
+            1
+        ],
+        [load_data_failure: InternalGas, "load_data.failure", 1],
+        // Gas parameters for writing data to storage.
+        [
+            write_data_per_op: InternalGasPerArg,
+            "write_data.per_op",
+            100
+        ],
+        [
+            write_data_per_new_item: InternalGasPerArg,
+            "write_data.new_item",
+            1000
+        ],
+        [
+            write_data_per_byte_in_key: InternalGasPerByte,
+            "write_data.per_byte_in_key",
+            100
+        ],
+        [
+            write_data_per_byte_in_val: InternalGasPerByte,
+            "write_data.per_byte_in_val",
+            100
+        ],
+        // Multiplier for write operations.
+        [
+            write_data_multiplier_max: InternalGasMultiplier,
+            "write_data.multiplier.max",
+            10
+        ],
+        [
+            write_data_multiplier_exp_base: InternalGasMultiplier,
+            "write_data.multiplier.exp_base",
+            32
+        ],
+        [
+            write_data_multiplier_target_items: NumArgs,
+            "write_data.multiplier.target.items",
+            1_000_000,
+        ],
+        [
+            write_data_multiplier_target_bytes: NumBytes,
+            "write_data.multiplier.target.bytes",
+            256 * 1024 * 1024 * 1024, /* 256 GB */
+        ],
     ]
 );
 
+fn calculate_write_multiplier<U>(
+    min: InternalGasMultiplier,
+    max: InternalGasMultiplier,
+    base: InternalGasMultiplier,
+    used: GasQuantity<U>,
+    target: GasQuantity<U>,
+) -> f64 {
+    // Clamp the params.
+    let min = cmp::max(min, 1.into());
+    let max = cmp::max(min, max);
+    let base = cmp::max(base, 1.into());
+
+    let target = cmp::max(target, 1.into());
+
+    // Convert to f64.
+    let min = u64::from(min) as f64;
+    let max = u64::from(max) as f64;
+    let base = u64::from(base) as f64;
+
+    let used = u64::from(used) as f64;
+    let target = u64::from(target) as f64;
+
+    if used >= target {
+        max
+    } else {
+        min + (max - min) * ((base.powf(used / target) - 1.0) / (base - 1.0))
```

Comment:
I'm concerned that we are doing a lot of floating point operations here.
1. Despite errors being rare today, I'm not sure if we can assume that all hardware our validator node will be running on will implement the IEEE 754 standard correctly. Remember the [Pentium FDIV bug](https://en.wikipedia.org/wiki/Pentium_FDIV_bug)
2. "Neither the IEEE 754 nor the C++ standard specify how transcendental functions like sin(), cos(), exp(), etc. have to be implemented. Terefore, for the same arguments, the results that these functions return on various platforms might not be exactly the same." https://www.appinf.com/download/FPIssues.pdf
3. Floating point operations are not associative. This makes the formula very fragile. Things could go wrong if you turn on compiler flags like ffast-math.

Reply:
Alternatively I wonder if we should just implement a tiered system that approximates @msmouse's curve with a few course steps.

Reply:
10 key points on chain and linear projection between the key points works for me.. 

figuring out if there's more elegant ways

Reply:
To be fair "writing the current price down on chain" mitigates this concern since once the pricing is agreed by all validators, they are integers on-chain, and replaying the transaction won't require a same result on the formula. 

-- we just need a quorum of validators to have the same result to keep the chain going and not worry about it later.

Reply:
@vgao1996 ideas?

Reply:
We can use Taylor series as an approximation of pow.

a^x = \sum_{i=0}^inf \frac{(xt)^i}{i!}, where t = lna

If we can make t to be integer, and we use integer division, the result should be deterministic.

Reply:
^ not sure if we see that unnecessarily computationally and conceptually heavy -- still trying to make my argument of writing down the actual per slot and per byte pricing in the gas schedule -- and if a 10 entry lookup table should be preferred.